module/zfs/arc.c

```diff
@@ -3374,14 +3378,22 @@ arc_convert_to_raw(arc_buf_t *buf, uint64_t dsobj, boolean_t byteorder,
     const uint8_t *mac)
 {
 	arc_buf_hdr_t *hdr = buf->b_hdr;
+	boolean_t need_transfer = B_TRUE;
 
 	ASSERT(ot == DMU_OT_DNODE || ot == DMU_OT_OBJSET);
 	ASSERT(HDR_HAS_L1HDR(hdr));
 	ASSERT3P(hdr->b_l1hdr.b_state, ==, arc_anon);
 
+	if (arc_buf_is_shared(buf)) {
+		arc_unshare_buf(hdr, buf);
+		arc_hdr_alloc_abd(hdr, B_FALSE);
+		abd_copy_from_buf(hdr->b_l1hdr.b_pabd, buf->b_data,
+		    HDR_GET_PSIZE(hdr));
+		need_transfer = B_FALSE;
+	}
```

Comment:
I would add an assert here for `!ARC_BUF_ENCRYPTED()`. the buf cant be encrypted because encrypted buffers can't be shared.